invenio_access/permissions.py

```diff
@@ -117,64 +118,78 @@ def _cache_key(action_need):
         """Helper method to generate cache key."""
         return get_action_cache_key(
             action_need.value,
-            action_need.argument if hasattr(action_need, 'argument') else None
+            action_need.argument if hasattr(action_need, "argument") else None,
         )
 
+    @staticmethod
+    def _split_actionsneeds(needs):
+        """Split needs into sets of ActionNeed and any other *Need."""
+        action_needs, other_needs = set(), set()
+        for need in needs:
+            action_needs.add(need) if need.method == "action" else \
+                other_needs.add(need)
+        return action_needs, other_needs
+
     def _load_permissions(self):
-        """Load permissions associated to actions."""
+        """Load permissions for all needs expanding actions."""
         result = _P(needs=set(), excludes=set())
-        if not self.allow_by_default:
-            result.needs.update(self.explicit_needs)
-            result.excludes.update(self.explicit_excludes)
-
-        for explicit_need in self.explicit_needs:
-            if explicit_need.method == 'action':
-                # in-place update of results
-                result.update(self._load_action(explicit_need))
-            elif self.allow_by_default:
-                result.needs.add(explicit_need)
-
-        for explicit_exclude in self.explicit_excludes:
-            if explicit_exclude.method == 'action':
-                # in-place update of results
-                result.update(self._load_action(explicit_exclude))
-            elif self.allow_by_default:
-                result.excludes.add(explicit_exclude)
+
+        # split ActionNeeds and any other *Needs in separates Sets
+        action_needs, explicit_needs = self._split_actionsneeds(
+            self.explicit_needs)
+        action_excludes, explicit_excludes = self._split_actionsneeds(
+            self.explicit_excludes)
+
+        # add all explicit needs/excludes to the result permissions
+        result.needs.update(explicit_needs)
+        result.excludes.update(explicit_excludes)
+
+        # expand all ActionNeeds to get all needs/excludes and add them to the
+        # result permissions
+        for need in action_needs | action_excludes:
+            result.update(self._expand_action(need))
+
+        # "allow_by_default = False" means that when needs/excludes are empty,
+        # then it should deny access.
+        # By default, `flask_principal.Permission.allows` will allow access
+        # if needs/excludes are empty!
+        needs_empty = len(result.needs) == 0 or len(result.excludes) == 0
```

Comment:
Question: Behavior will be the same, since we are trying to `forbid`. But checking ony `len(result.needs)  == 0`  should be enough since the case is that the `need` is not assign to anybody (therefore not expanded). There's no need to check for excludes (and `update` them below). 

However, for consistency/easiness of understanding, it might still be good to check and add? This was mainly I question to see if I understood it correctly.

Reply:
Removed as discussed.